ai/WFLY-19381-[EXPERIMENTAL]-Provide_a_galleon_feature_pack_to_facilitate_Genrative_AI_development.adoc

```diff
@@ -0,0 +1,264 @@
+ * content retrievers (aka retrievers of content to provide to the llm as part of the context based on the user query): `dev.langchain4j.rag.content.retriever.ContentRetriever`
+ * chat language models (aka a chat APi with the llm): `dev.langchain4j.model.chat.ChatLanguageModel`
+ Those resources should be exposed via CDI (and thus Weld) to the application using qualifier and type.
+ The less WildFly specific annotations are used the better so this feature should try to use annotations from librairies that are already used in WildFly like smallrye-common-annotations or annotations from LangChain4J.
+ We should provide layers to provision the server according to the needs.
+
+
+=== Nice-to-Have Requirements
+
+ * Provide annotations to be able to create AIServices using our configured resources.
+ * Replace the HTTP clients used so that we have only one that is supported for every API.
+ * Replace the JSON marshalling/umarshalling librairies so that we have only one (through RESTEasy) that is supported.
+ * Support for @Tool
+ * Adding support for ChatMemory
+ * Adding support for more APIs
+
+=== Non-Requirements
+// Use this section to explicitly discuss things that readers might think are required
+// but which are not required.
+
+=== Future Work
+// Use this section to discuss requirements that are not addressed by this proposal
+// but which may be addressed in later proposals.
+
+== Backwards Compatibility
+
+// Does this enhancement affect backwards compatibility with previously released
+// versions of WildFly?
+// Can the identified incompatibility be avoided?
+
+=== Default Configuration
+
+=== Importing Existing Configuration
+
+=== Deployments
+
+The required librairies should be added automatically on the deployment classpath. 
+
+=== Interoperability
+
+== Implementation Plan
+
+=== Embeddings models
+
+The extension should provide resources to define `dev.langchain4j.model.embedding.EmbeddingModel`.
+
+It should expose a simple `embedding-model` resource with the following attributes:
+
+* module: the jboss module containing the code of the model.
+* embedding-class: the name of the class to use to load the model.
+
+----
+/subsystem=ai/embedding-model=myembedding:add(module=dev.langchain4j.embeddings.all-minilm-l6-v2, embedding-class=dev.langchain4j.model.embedding.AllMiniLmL6V2EmbeddingModel)
+----
+
+It should also support LLM backed embedding models like for ollama for example.
+We should have an `ollama-embedding-model` resource with the following attributes:
+
+* base-url: endpoint to connect to an Ollama chat model.
+* connect-timeout: timeout for the Ollama embedding model.
+* log-requests: enabling the tracing of requests going to Ollama.
+* log-responses: enabling the tracing of responses from Ollama.
+* model-name: the name of the embedding model served by Ollama.
+
+----
+subsystem=ai/ollama-embedding-model=myembedding:add(base-url="http://192.168.1.11:11434", model-name="llama3:8b")
```

Comment:
I will pick `ollama-embedding-model`, but this is general objection.  I am asking myself how good idea is to map LangChain4J model API to Wildfly model 1:1. LangChain4j API is  very broad and will change.  Wildfly model will always be lagging. Aren't we misusing here wildfly model as Inversion of Control system here?

This is different to Hibernate (jpa) subsystem or datasource (jdbc) subsystem which are small. This `ai` subsystem can be over time very big (hard to maintain) subsystem.

Reply:
Well, I wondered myself. langchain4j is already an abstraction over the llms and other elements of the rag chain so adding one more is maybe too much. Also you don't have to match 1-1, you only need to expose those attributes you think make sense. Would having a map of attributes to a generic type that would somehow do the matching be more to your liking ?

Reply:
This basic topic seems like the key differentiator between an experimental stability offering and preview.